### ISSUE TITLE
Requests fail after hot code upgrade

### ISSUE THREAD

**Issue by dmarkow** [benoitc/hackney#492]
Since 98f5ac60b1e67a5fbbb3fa9b002ff699ca429c97 was merged (which moved the `metrics:init` call into start/2), a hot code upgrade of my app results in failed hackney calls until I fully restart my elixir application.

> 2018-03-22 07:07:36.840  [error] Ranch listener ApiWeb.Endpoint.HTTP had connection process started with :cowboy_protocol:start_link/4 at #PID<0.3108.0> exit with reason: {{:badmatch, :undefined}, [{:hackney_connect, :create_connection, 5, [file: '/data/www/pfcs-api-builds/deps/hackney/src/hackney_connect.erl', line: 69]}, {:hackney_connect, :connect, 5, [file: '/data/www/pfcs-api-builds/deps/hackney/src/hackney_connect.erl', line: 37]}, {:hackney, :request, 5, [file: '/data/www/pfcs-api-builds/deps/hackney/src/hackney.erl', line: 315]}, {ExAws.Request.Hackney, :request, 5, [file: 'lib/ex_aws/request/hackney.ex', line: 20]}, {ExAws.Request, :request_and_retry, 7, [file: 'lib/ex_aws/request.ex', line: 38]}, {ExAws.Operation.ExAws.Operation.S3, :perform, 2, [file: 'lib/ex_aws/operation/s3.ex', line: 40]}, {ExAws.S3.Upload, :initialize, 2, [file: 'lib/ex_aws/s3/upload.ex', line: 40]}, {ExAws.Operation.ExAws.S3.Upload, :perform, 2, [file: 'lib/ex_aws/s3/upload.ex', line: 82]}]}

I traced this back to following line in hackney_connect.erl returning `undefined` after the reload:

```
{ok, Engine} = application:get_env(hackney, metrics),
```

I believe this is because the new `set_env` call is reset by the app config in hackney.app.src when it's reloaded, removing the `metrics` option?

**Comment by redink** [comment]
ME TOO.

before upgrade, I can get all env of hackney:

```
iex(exam@127.0.0.1)2> Application.get_all_env(:hackney)
[
  maxr: 10,
  shutdown: 10000,
  included_applications: [],
  metrics: {:metrics_ng, :metrics_dummy},
  maxt: 1,
  restart: :permanent,
  max_connections: 50,
  timeout: 150000
]
```

but, after upgrade, the `metrics` env lost:

```
iex(exam@127.0.0.1)4> Application.get_all_env(:hackney)
[
  maxr: 10,
  shutdown: 10000,
  included_applications: [],
  maxt: 1,
  restart: :permanent,
  max_connections: 50,
  timeout: 150000
]
```

**Comment by benoitc** [comment]
I'm not sure about the change in the PR, [the module is created when the `hackney` application is started ](https://github.com/benoitc/hackney/blob/master/src/hackney_app.erl#L23 )to make sure it's already compiled. 

You're right however this env will be reset. I propose then to compile it and maintain it in it's own ets instead so we can be sure it will be kept..

**Comment by benoitc** [comment]
fixed in 9604828584457724d1502a535edc8cc8b0f43b94 . Thanks for the initial patch!
